File: leetcode/python/2353_design_food_rating_system.py
import heapq


class FoodRatings:
    def __init__(self, foods: list[str], cuisines: list[str], ratings: list[int]):
        self.cuisine_heap = {}
        self.rating = {}
        self.food_to_cuisine = {}
        
        for food, cuisine, rating in zip(foods, cuisines, ratings):
            if cuisine not in self.cuisine_heap:
                self.cuisine_heap[cuisine] = []
            self.cuisine_heap[cuisine].append((-rating, food))
            self.rating[food] = rating
            self.food_to_cuisine[food] = cuisine
            
        for cuisine in self.cuisine_heap:
            heapq.heapify(self.cuisine_heap[cuisine])
            
    def changeRating(self, food: str, newRating: int) -> None:
        cuisine = self.food_to_cuisine[food]
        self.rating[food] = newRating
        heapq.heappush(self.cuisine_heap[cuisine], (-newRating, food))

    def highestRated(self, cuisine: str) -> str:
        while self.cuisine_heap[cuisine]:
            rating, food = self.cuisine_heap[cuisine][0]
            if rating == -self.rating[food]:
                return food
            else:
                heapq.heappop(self.cuisine_heap[cuisine])
        return ''
        

File: leetcode/rust/src/2353_design_food_rating_system.rs
use std::{collections::{HashMap, BinaryHeap}, cmp::Reverse};

struct FoodRatings {
    heaps: HashMap<String, BinaryHeap<(i32, Reverse<String>)>>,
    ratings: HashMap<String, i32>,
    cuisines: HashMap<String, String>,
}

/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl FoodRatings {

    fn new(foods: Vec<String>, cuisines: Vec<String>, ratings: Vec<i32>) -> Self {
        let mut inst = Self {
            heaps: HashMap::new(),
            ratings: HashMap::new(),
            cuisines: HashMap::new(),
        };

        for i in 0..foods.len() {
            let food = foods[i].clone();
            let cuisine = cuisines[i].clone();
            let rating = ratings[i];
            if !inst.heaps.contains_key(&cuisine) {
                inst.heaps.insert(cuisine.clone(), BinaryHeap::new());
            }

            inst.heaps
                .get_mut(&cuisine).unwrap()
                .push((rating, Reverse(food.clone())));
            inst.ratings.insert(food.clone(), rating);
            inst.cuisines.insert(food.clone(), cuisine);
        }
        inst
    }
    
    fn change_rating(&mut self, food: String, new_rating: i32) {
        self.ratings.insert(food.clone(), new_rating);
        let cuisine = &self.cuisines[&food];
        self.heaps
            .get_mut(cuisine).unwrap()
            .push((new_rating, Reverse(food.clone())));

        // self.heaps[cuisine].push((new_rating, food.clone()));
    }
    
    fn highest_rated(&mut self, cuisine: String) -> String {
        while !self.heaps[&cuisine].is_empty() {
            let (rating, Reverse(food)) = self.heaps[&cuisine].peek().unwrap();
            if self.ratings[food] == *rating {
                return food.clone();
            } else {
                self.heaps.get_mut(&cuisine).unwrap().pop();
            }
        }

        "".to_string()
    }
}
